src/fqs/fqs.py

```diff
@@ -0,0 +1,375 @@
+    '''
+    with tf.name_scope("linear"):
+        # n.b. this will give nonsense solutions and really should return NaN at a0 because there is no solution
+        (_, a0) = _safe_denom(a0)
+        return -b0 / a0
+
+
+@tf.function(experimental_autograph_options=tf.autograph.experimental.Feature.NAME_SCOPES)
+def quadratic(a0: tf.Tensor,
+              b0: tf.Tensor,
+              c0: tf.Tensor,
+              assume_quadratic=True,
+              **kwargs) -> Tuple[tf.Tensor, tf.Tensor]:
+    ''' Analytical solver for a single quadratic equation
+    (2nd order polynomial).
+
+    Parameters
+    ----------
+    a0, b0, c0: array_like
+        Input data are coefficients of the Quadratic polynomial::
+
+            a0*x^2 + b0*x + c0 = 0
+
+    the `a0` elements must all be non-zero
+
+    Returns
+    -------
+    r1, r2: tuple
+        Output data is a tuple of two roots of a given polynomial.
+    '''
+    with tf.name_scope("quadratic"):
+        if not assume_quadratic:
+            (flag_degenerate, a0) = _safe_denom(a0)
+        # Reduce the quadratic equation to the form:
+        #    x^2 + bx + c = 0
+        b, c = b0 / a0, c0 / a0
+
+        # Some repeating variables
+        n_half_b = -0.5 * b
+        delta = n_half_b * n_half_b - c
+        sqrt_delta = square_root(ensure_complex(delta))
+        n_half_b = ensure_complex(n_half_b)
+
+        # Roots
+        r1 = n_half_b - sqrt_delta
+        r2 = n_half_b + sqrt_delta
+
+        if assume_quadratic:
+            return r1, r2
+        else:
+            # linear fall-back
+            with tf.name_scope("linear_fallback"):
+                flag_degenerate = (a0 == 0)
+                r0 = ensure_complex(linear(b0, c0))
+                return tf.where(flag_degenerate, r0, r1), tf.where(flag_degenerate, r0, r2)
+
+
+@tf.function(experimental_autograph_options=tf.autograph.experimental.Feature.NAME_SCOPES)
+def cubic(a0: tf.Tensor,
+          b0: tf.Tensor,
+          c0: tf.Tensor,
+          d0: tf.Tensor,
+          first_root_only: bool = False,
+          assume_cubic: bool = True,
+          **kwargs) -> Union[Tuple[tf.Tensor, tf.Tensor, tf.Tensor], Tuple[tf.Tensor]]:
+    ''' Analytical closed-form solver for a single cubic equation
+    (3rd order polynomial), gives all three roots.
+
+    Parameters
+    ----------
+    a0, b0, c0, d0: array_like
+        Input data are coefficients of the Cubic polynomial::
+
+            a0*x^3 + b0*x^2 + c0*x + d0 = 0
+
+    the `a0` elements must all be non-zero
+
+    Returns
+    -------
+    roots: tuple
+        Output data is a tuple of three roots of a given polynomial.
+    '''
+    with tf.name_scope("cubic"):
+        if not assume_cubic:
+            (quadratic_mask, a0) = _safe_denom(a0)
+        #if unique_real_root and (a0.dtype.is_complex or b0.dtype.is_complex or c0.dtype.is_complex or d0.dtype.is_complex):
+        #    raise TypeError("There may be no real root if the coefficients are complex")
```

Comment:
```suggestion
```

Reply:
Addressed in [4b94568](https://github.com/Geopipe/fqs/pull/1/commits/4b94568e86e1d7ca1f66a29dffb5a0a867b8bb95)